test/object-store/c_api/c_api.cpp

```diff
@@ -2328,6 +2328,126 @@ TEST_CASE("C API") {
+            CHECK(realm_is_frozen(frozen_realm.get()));
+            CHECK(!realm_is_frozen(thawed_realm.get()));
+        }
+
+        SECTION("objects") {
+            CPtr<realm_object_t> obj1;
+            realm_value_t value;
+
+            write([&]() {
+                obj1 = cptr_checked(realm_object_create(realm, class_foo.key));
+                CHECK(obj1);
+            });
+            CHECK(checked(realm_get_value(obj1.get(), foo_str_key, &value)));
+            CHECK(value.type == RLM_TYPE_STRING);
+            CHECK(strncmp(value.string.data, "", value.string.size) == 0);
+
+            auto frozen_realm = cptr_checked(realm_freeze(realm));
+            auto frozen_obj1 = cptr_checked(realm_object_freeze(obj1.get(), frozen_realm.get()));
+            CHECK(frozen_obj1);
+
+            write([&]() {
+                CHECK(checked(realm_set_value(obj1.get(), foo_str_key, rlm_str_val("Hello, World!"), false)));
+            });
+
+            CHECK(checked(realm_get_value(obj1.get(), foo_str_key, &value)));
+            CHECK(value.type == RLM_TYPE_STRING);
+            CHECK(strncmp(value.string.data, "Hello, World!", value.string.size) == 0);
+
+            CHECK(checked(realm_get_value(frozen_obj1.get(), foo_str_key, &value)));
+            CHECK(value.type == RLM_TYPE_STRING);
+            CHECK(strncmp(value.string.data, "", value.string.size) == 0);
+
+            auto thawed_obj1 = cptr_checked(realm_object_thaw(frozen_obj1.get(), realm));
+            CHECK(thawed_obj1);
+            CHECK(checked(realm_get_value(thawed_obj1.get(), foo_str_key, &value)));
+            CHECK(value.type == RLM_TYPE_STRING);
+            CHECK(strncmp(value.string.data, "Hello, World!", value.string.size) == 0);
+        }
+
+        SECTION("results") {
+            auto results = cptr_checked(realm_object_find_all(realm, class_foo.key));
+            realm_results_delete_all(results.get());
+
+            write([&]() {
+                // Ensure that we start from a known initial state
+                CHECK(realm_results_delete_all(results.get()));
+
+                auto obj1 = cptr_checked(realm_object_create(realm, class_foo.key));
+                CHECK(obj1);
+            });
+
+            size_t count;
+            realm_results_count(results.get(), &count);
+            CHECK(count == 1);
+
+            auto frozen_realm = cptr_checked(realm_freeze(realm));
+            auto frozen_results = cptr_checked(realm_results_freeze(results.get(), frozen_realm.get()));
+            write([&]() {
+                auto obj1 = cptr_checked(realm_object_create(realm, class_foo.key));
+                CHECK(obj1);
+            });
+            realm_results_count(frozen_results.get(), &count);
+            CHECK(count == 1);
+            realm_results_count(results.get(), &count);
+            CHECK(count == 2);
+
+            // TODO Freezing a result with a live realm crashes due to an assert. Is that fine or
```

Comment:
I think in general in ObjectStore we use asserts to enforce pre-conditions before calling an API, but I don't have strong opinions on this.